Implement GPT-image-1 token-based cost tracking

colesmcintosh opened the pull request:
## Title

Implement GPT-image-1 token-based cost tracking

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🆕 New Feature

## Changes

This PR implements token-based cost tracking for GPT-image-1 models, replacing the previous pixel-based pricing approach with OpenAI's actual token-based pricing model.

### Key Changes:

**1. Pricing Configuration (`model_prices_and_context_window.json`)**
- Updated GPT-image-1 variants to use token-based pricing:
  - Text input: $5/1M tokens (`input_cost_per_token`: 5e-06)
  - Image input: $10/1M tokens (`input_cost_per_image_token`: 1e-05) 
  - Image output: $40/1M tokens (`output_cost_per_token`: 4e-05)
- Applied across all quality levels (low: 272 tokens, medium: 1,056 tokens, high: 4,160 tokens)

**2. Core Implementation (`litellm/cost_calculator.py`)**
- Added `_estimate_gpt_image_1_output_tokens()` function for token estimation based on quality/dimensions
- Enhanced `default_image_cost_calculator()` to use actual token usage when available
- Updated `completion_cost()` integration to pass real token counts
- Maintained fallback estimation for cases without usage data

**3. Comprehensive Test Suite**
- Added comprehensive tests covering token estimation, cost calculation, and integration
- Tests validate different quality levels, image sizes, Azure vs OpenAI variants
- Integration tests with existing cost calculation framework
- Edge cases and fallback scenario validation

### Token Estimation Logic:
- **Low quality**: 272 tokens per image
- **Medium quality**: 1,056 tokens per image  
- **High quality**: 4,160 tokens per image
- Dynamic calculation based on image dimensions and quality settings

### Test Results:
All tests pass - 2 GPT-image-1 specific tests added and passing.

**Screenshot of test results:** 
<img width="773" alt="Screenshot 2025-06-08 at 3 41 27 PM" src="https://github.com/user-attachments/assets/ba63b2ef-6e70-43f8-948b-72f7cc7ea6ee" />

**Files Changed:**
- `model_prices_and_context_window.json` - Pricing configuration
- `litellm/cost_calculator.py` - Core implementation
- `tests/test_litellm/test_cost_calculator.py` - Integration test
- `tests/local_testing/test_completion_cost.py` - Completion cost test

colesmcintosh committed:
feat: update GPT-image-1 pricing to token-based model - Replace pixel-based pricing with OpenAI token-based pricing - Add input_cost_per_token (5e-06), output_cost_per_token (4e-05), input_cost_per_image_token (1e-05) - Update all GPT-image-1 variants (OpenAI and Azure) across all quality levels

colesmcintosh committed:
feat: implement GPT-image-1 token-based cost calculation - Add token estimation function based on quality and dimensions - Modify default_image_cost_calculator to use actual token usage when available - Update completion_cost to pass real token usage to image cost calculator - Maintain fallback estimation for cases without usage data

colesmcintosh committed:
test: add comprehensive tests for GPT-image-1 cost tracking - Add token estimation validation tests - Add cost calculation accuracy tests for all quality levels - Add integration tests with completion_cost function - Test fallback estimation and edge cases - Validate correct cost differences across quality levels - Ensures accurate token-based pricing implementation

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
test: add input_cost_per_image_token to validation schema for GPT-image-1 pricing

ishaan-jaff commented on litellm/cost_calculator.py:
```diff
@@ -1155,7 +1159,74 @@ def default_image_cost_calculator(
-    return cost_info["input_cost_per_pixel"] * height * width * n
+    # Check if this is a token-based pricing model (like gpt-image-1)
+    if "input_cost_per_token" in cost_info or "output_cost_per_token" in cost_info:
+        # For token-based pricing, we need to calculate based on actual usage
```
this should be a helper func

ishaan-jaff commented on litellm/cost_calculator.py:
```diff
@@ -1155,7 +1159,74 @@ def default_image_cost_calculator(
+        # Real cost should be calculated from actual token usage in response
+        
+        # For gpt-image-1, estimate output tokens based on quality and size
+        base_model_str = model.split("/")[-1] if "/" in model else model
```
would this work for gpt-image on azure? the user uses azure gpt-image

ishaan-jaff reviewed (changes requested):
LGTM, minor changes. Please can you verify this works for Azure gpt image

colesmcintosh committed:
feat: refactor GPT-image-1 cost calculation to use dedicated function for token-based pricing - Implement _calculate_token_based_image_cost to streamline cost computation - Update default_image_cost_calculator to leverage new function - Enhance tests to validate cost calculations with actual token usage for various quality levels and Azure deployment names

colesmcintosh committed:
fix: ensure n is never None in default_image_cost_calculator - Introduce n_value to default to 1 when n is None, ensuring accurate cost calculations for both token-based and pixel-based pricing models

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
Enhance cost calculator for Azure custom deployments and improve token-based pricing handling. Added fallback for Azure model names and refined cost calculation logic for token-based models. Updated documentation for clarity on token requirements and backward compatibility with existing tests.

colesmcintosh committed:
Refactor default_image_cost_calculator to ensure size_str is always a concrete string and standardize dimension delimiter to "-x-". This improves type safety and consistency in size formatting.

colesmcintosh committed:
Enhance default_image_cost_calculator by adding checks for None values before accessing model costs. This improves type safety and ensures that only valid models are processed.

colesmcintosh committed:
Refactor default_image_cost_calculator to improve type safety by asserting non-None values for custom_llm_provider and explicitly defining model name variables. This enhances clarity and ensures valid model names are processed for Azure deployments.

colesmcintosh committed:
Refactor default_image_cost_calculator to ensure custom_llm_provider is explicitly checked for None before processing model names. This improves type safety and clarity in model name handling.

colesmcintosh committed:
Merge main branch into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
Refactor cost calculator: rename loop variable for clarity in default_image_cost_calculator function

krrishdholakia commented on tests/local_testing/test_completion_cost.py:
```diff
@@ -319,6 +319,80 @@ def test_cost_openai_image_gen():
     assert cost == 0.019922944
 
 
+def test_cost_gpt_image_1_token_based():
```
move to `tests_litellm/` as this seems like a mock test as well

krrishdholakia commented on tests/local_testing/test_completion_cost.py:
```diff
@@ -319,6 +319,80 @@ def test_cost_openai_image_gen():
+    high_response.usage = Usage(prompt_tokens=50, completion_tokens=4160, total_tokens=4210)
+    
+    # Test different quality levels with actual response data
+    cost_low = litellm.completion_cost(
```
add try-excepts around this - so if the tests fail due to openai internal server errors / rate limit errors - it doesn't stop ci/cd

krrishdholakia commented on tests/local_testing/test_completion_cost.py:
```diff
@@ -319,6 +319,80 @@ def test_cost_openai_image_gen():
+        call_type="image_generation",
+    )
+    
+    cost_high = litellm.completion_cost(
```
can we avoid calling the model 3 times in 1 test - it'll get pretty expensive. 1 real call is fine

krrishdholakia commented on litellm/cost_calculator.py:
```diff
@@ -1151,16 +1154,129 @@ def default_image_cost_calculator(
+    n_value = n if n is not None else 1
+
+    # Decide pricing type
+    if "input_cost_per_token" in cost_info or "output_cost_per_token" in cost_info:
```
aren't these guaranteed fields? 

the check should be if their non-null

krrishdholakia replied:
nvm - not guaranteed if you use `litellm._model_cost`, guaranteed if you use `get_model_info`

krrishdholakia commented on litellm/cost_calculator.py:
```diff
@@ -1151,16 +1154,129 @@ def default_image_cost_calculator(
+
+    # Validate model family – we only support gpt-image-1-style pricing here so
+    # fail fast for others (unit-tests rely on this behaviour).
+    if "gpt-image-1" not in model:
```
why hardcode this? 

as long as the cost map has input cost per token / output cost per token pricing info, this should not matter

krrishdholakia commented on litellm/cost_calculator.py:
```diff
@@ -1151,16 +1154,129 @@ def default_image_cost_calculator(
+    return (input_cost + output_cost) * n
+
+
+def _estimate_gpt_image_1_output_tokens(
```
where is this used?

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
Refactor cost calculation for token-based image models in `_calculate_token_based_image_cost`. Improved validation for cost configuration and clarified function documentation.

colesmcintosh committed:
Update tests for token-based image cost calculation. Enhance error handling for missing token pricing configuration and clarify assertions in test cases.

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking

colesmcintosh committed:
feat: address PR review feedback for GPT-image-1 token-based cost tracking

- Extract token-based pricing logic into dedicated helper function
- Replace direct model_cost lookups with get_model_info() API
- Improve Azure support by removing hardcoded gpt-image-1 references
- Support multiple image models (gpt-image-1, dall-e-3, dall-e-2) for Azure
- Update tests to use mocks and handle API errors gracefully
- Fix type annotations to support both dict and ModelInfo objects
- Ensure all mypy type checks pass

Addresses feedback from @ishaan-jaff and @krrishdholakia

colesmcintosh committed:
fix: refactor default_image_cost_calculator to reduce statements

- Extract _build_model_name_variants() helper function
- Extract _find_azure_token_based_model() helper function
- Reduces function from 52 to under 50 statements
- Fixes PLR0915 linting error

colesmcintosh committed:
fix: update test_default_image_cost_calculator to mock get_model_info

- Test was failing because it relied on monkeypatching litellm.model_cost
- Our refactored code now uses get_model_info() which doesn't respect the patch
- Fixed by mocking get_model_info to return the test model data
- Ensures test works with the new implementation

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/lit-202-gpt-image-1-token-cost-tracking
